packages/klighd-core/src/filtering/semantic-filtering-util.ts

```diff
@@ -0,0 +1,318 @@
+ */
+export class IfThenConnective implements BinaryConnective {
+    static NAME = "IFTHEN"
+    name = IfThenConnective.NAME
+    leftOperand: SemanticFilterRule
+    rightOperand: SemanticFilterRule
+    ruleName?: string
+}
+
+/**
+ * A LogicEqual Connective takes two rules R1 and R2 and evaluates to true
+ * iff
+ * R1 and R2 evaluate to true or R1 and R2 evaluate to false.
+ * @example R1 === R2
+ * @example R1 && R2 || !R1 && !R2
+ */
+export class LogicEqualConnective implements BinaryConnective {
+    static NAME = "LOGICEQUAL"
+    name = LogicEqualConnective.NAME
+    leftOperand: SemanticFilterRule
+    rightOperand: SemanticFilterRule
+    ruleName?: string
+}
+
+/**
+ * An IfThenElse Connective takes three rules R1, R2 and R3 and evaluates to true
+ * iff
+ * R1 and R2 evaluate to true or R1 evaluates to false and R3 evaluates to true.
+ * @example R1 ? R2 : R3
+ * @example R1 && R2 || !R1 && R3
+ */
+export class IfThenElseConnective implements TernaryConnective {
+    static NAME = "IFTHENELSE"
+    name = IfThenElseConnective.NAME
+    firstOperand: SemanticFilterRule
+    secondOperand: SemanticFilterRule
+    thirdOperand: SemanticFilterRule
+    ruleName?: string
+}
+
+//// Numeric Connectives ////
+
+/**
+ * A LessThan Connective takes one rule R and evaluates to true
+ * iff
+ * R.num < correspondingTag.num.
+ * @example R.num < correspondingTag.num
+ */
+export class LessThanConnective implements UnaryConnective {
+    static NAME = "LESSTHAN"
+    name = LessThanConnective.NAME
+    operand: SemanticFilterTag
+    ruleName?: string
+}
+
+/**
+ * A GreaterThan Connective takes one rule R and evaluates to true
+ * iff
+ * R.num > correspondingTag.num.
+ * @example R.num > correspondingTag.num
+ */
+export class GreaterThanConnective implements UnaryConnective {
+    static NAME = "GREATERTHAN"
+    name = GreaterThanConnective.NAME
+    operand: SemanticFilterTag
+    ruleName?: string
+}
+
+/**
+ * A NumericEqual Connective takes one rule R and evaluates to true
+ * iff
+ * R.num === correspondingTag.num.
+ * @example R.num === correspondingTag.num
+ */
+export class NumericEqualConnective implements UnaryConnective {
+    static NAME = "NUMERICEQUAL"
+    name = NumericEqualConnective.NAME
+    operand: SemanticFilterTag
+    ruleName?: string
+}
+
+//// Functions ////
+
+/**
+ * A filter is used to apply a filter rule as a boolean function on a graph element. The function
+ * returns true if the element fulfils the filter rule and false otherwise.
+ */
+export interface Filter {
+    name?: string
+    defaultValue?: boolean
+    filterFun(el: SKGraphElement): boolean
+}
+
+/**
+ * Creates a new filter with a function that can be applied to graph elements when given 
+ * a filter rule.
+ * @param rule the rule to construct the filter from
+ * @returns a new filter
+ */
+export function createFilter(rule: SemanticFilterRule): Filter {
+
+    let ruleName;
+    if (rule instanceof SemanticFilterTag) {
```

Comment:
Does this `instanceof` check work for rules defined by the server and serialized/deserialized? Usually the resulting objects are plain JS objects with fields and no such working `instanceof`. Usually this requires some [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates)

Reply:
this works